lkcrypto.c

```diff
@@ -198,18 +272,115 @@ static int decrypt_encrypt(lua_State *L, int dec_enc)
 		memcpy(iv, iv_orig, iv_len);
 	}
 	buffer = kmalloc(ciphertext_len, GFP_ATOMIC);
+	if (buffer == NULL) {
+		return luaL_error(L, "buffer allocation failed");
+	}
 	memcpy(buffer, ciphertext, ciphertext_len);
 	result = enc_dec(tfm, key, key_len, iv, iv_len, buffer, ciphertext_len,
-			 dec_enc);
+			 operation);
 
 	if (!result) {
 		lua_pushlstring(L, buffer, ciphertext_len);
 		kfree(buffer);
 		return 1;
 	} else {
 		kfree(buffer);
-		return luaL_error(L, "%s failed %d",
-				  dec_enc ? "encrypt" : "decrypt");
+		return luaL_error(L, "%s failed",
+				  operation == ENCODING ? "encrypt" :
+							  "decrypt");
+	}
+}
+
+static int decrypt_encrypt_aead(lua_State *L, enum enc_dec operation)
+{
+	size_t key_len = 0;
+	size_t iv_len = 0;
+	size_t auth_len = 0;
+	size_t ciphertext_len = 0;
+	size_t aad_len = 0;
+
+	int result = 0;
+	int nargs = 0;
+	int aad_arg = operation == ENCODING ? 5 : 6;
+	const u8 *key;
+	const u8 *iv_orig;
+	const u8 *auth_orig;
+	const u8 *aad_orig;
+	const u8 *ciphertext;
+	u8 iv[AES_MAX_KEY_SIZE] = {
+		0,
+	};
+	u8 *auth = NULL;
+	u8 *buffer = NULL;
+	u8 *aad = NULL;
+
+	struct cipher_aead_tfm_t *tfm = NULL;
+
+	nargs = lua_gettop(L);
+
+	tfm = luaL_checkudata(L, 1, CUJO_CIPHER_AEAD_TFM_LUA_NAME);
+	key = luaL_checklstring(L, 2, &key_len);
+	ciphertext = luaL_checklstring(L, 3, &ciphertext_len);
+	iv_orig = luaL_checklstring(L, 4, &iv_len);
+	memcpy(iv, iv_orig, iv_len);
+	if (operation == DECODING) {
+		auth_orig = luaL_checklstring(L, 5, &auth_len);
+		auth = kmalloc(auth_len, GFP_ATOMIC);
+		if (auth == NULL) {
+			return luaL_error(L, "auth allocation failed");
+		}
+		memcpy(auth, auth_orig, auth_len);
+	}
+
+	else {
```

Comment:
Well, with `else` it should be on the same line as the `}`...

Reply:
addressed

Reply:
regarding the comment thing - I haven't investigate fully what is actually happening there